File: conversation-management/csharp/ConversationManagement.cs
/**
 * SendSeven API - Conversation Management Example (C#)
 *
 * Demonstrates how to list, get, update, and close conversations using the SendSeven API.
 */

using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Web;

class Program
{
    private static readonly HttpClient httpClient = new HttpClient();

    // Configuration from environment
    private static string ApiToken => Environment.GetEnvironmentVariable("SENDSEVEN_API_TOKEN") ?? "";
    private static string TenantId => Environment.GetEnvironmentVariable("SENDSEVEN_TENANT_ID") ?? "";
    private static string ApiUrl => Environment.GetEnvironmentVariable("SENDSEVEN_API_URL") ?? "https://api.sendseven.com/api/v1";

    static async Task Main(string[] args)
    {
        // Load .env file if it exists
        LoadEnvFile();

        // Validate configuration
        if (string.IsNullOrEmpty(ApiToken))
        {
            Console.WriteLine("Error: SENDSEVEN_API_TOKEN environment variable is required");
            Environment.Exit(1);
        }

        if (string.IsNullOrEmpty(TenantId))
        {
            Console.WriteLine("Error: SENDSEVEN_TENANT_ID environment variable is required");
            Environment.Exit(1);
        }

        try
        {
            // Example 1: List all open conversations that need a reply
            PrintSeparator();
            Console.WriteLine("Listing open conversations that need a reply...");
            PrintSeparator();

            var result = await ListConversationsAsync(new Dictionary<string, string>
            {
                { "status", "open" },
                { "needs_reply", "true" },
                { "page_size", "5" }
            });

            var pagination = result.GetProperty("pagination");
            Console.WriteLine($"Found {pagination.GetProperty("total").GetInt32()} conversations");
            Console.WriteLine($"Page {pagination.GetProperty("page").GetInt32()} of {pagination.GetProperty("total_pages").GetInt32()}");
            Console.WriteLine();

            var items = result.GetProperty("items");
            foreach (var conv in items.EnumerateArray())
            {
                Console.WriteLine($"  ID: {conv.GetProperty("id").GetString()}");
                Console.WriteLine($"  Channel: {conv.GetProperty("channel").GetString()}");
                Console.WriteLine($"  Status: {conv.GetProperty("status").GetString()}");
                var lastMsg = conv.TryGetProperty("last_message_at", out var lm) && lm.ValueKind != JsonValueKind.Null
                    ? lm.GetString() : "N/A";
                Console.WriteLine($"  Last message: {lastMsg}");
                Console.WriteLine();
            }

            // Example 2: Get a single conversation (if we have any)
            if (items.GetArrayLength() > 0)
            {
                var conversationId = items[0].GetProperty("id").GetString()!;

                PrintSeparator();
                Console.WriteLine($"Getting conversation details: {conversationId}");
                PrintSeparator();

                var conversation = await GetConversationAsync(conversationId);
                Console.WriteLine($"  ID: {conversation.GetProperty("id").GetString()}");
                Console.WriteLine($"  Channel: {conversation.GetProperty("channel").GetString()}");
                Console.WriteLine($"  Status: {conversation.GetProperty("status").GetString()}");
                Console.WriteLine($"  Needs reply: {conversation.GetProperty("needs_reply").GetBoolean()}");
                var assignedTo = conversation.TryGetProperty("assigned_to", out var at) && at.ValueKind != JsonValueKind.Null
                    ? at.GetString() : "Unassigned";
                Console.WriteLine($"  Assigned to: {assignedTo}");
                if (conversation.TryGetProperty("contact", out var contact) && contact.ValueKind != JsonValueKind.Null)
                {
                    var name = contact.TryGetProperty("name", out var n) && n.ValueKind != JsonValueKind.Null
                        ? n.GetString() : "Unknown";
                    Console.WriteLine($"  Contact: {name}");
                }
                Console.WriteLine();

                // Example 3: Demonstrate update (commented out to avoid modifying data)
                // Uncomment to actually assign a conversation
                // PrintSeparator();
                // Console.WriteLine("Assigning conversation to user...");
                // PrintSeparator();
                // var userId = "your-user-id-here";
                // var updated = await UpdateConversationAsync(conversationId, userId);
                // Console.WriteLine($"  Assigned to: {updated.GetProperty("assigned_to").GetString()}");
                // Console.WriteLine();

                // Example 4: Demonstrate close (commented out to avoid modifying data)
                // Uncomment to actually close the conversation
                // PrintSeparator();
                // Console.WriteLine("Closing conversation...");
                // PrintSeparator();
                // var closed = await CloseConversationAsync(conversationId);
                // Console.WriteLine($"  Status: {closed.GetProperty("status").GetString()}");
                // Console.WriteLine($"  Closed at: {closed.GetProperty("closed_at").GetString()}");
            }

            PrintSeparator();
            Console.WriteLine("Conversation management examples completed!");
            PrintSeparator();
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error: {ex.Message}");
            Environment.Exit(1);
        }
    }

    /// <summary>
    /// List conversations with optional filtering.
    /// </summary>
    private static async Task<JsonElement> ListConversationsAsync(Dictionary<string, string> options)
    {
        var queryParams = HttpUtility.ParseQueryString(string.Empty);
        queryParams["page"] = options.GetValueOrDefault("page", "1");
        queryParams["page_size"] = options.GetValueOrDefault("page_size", "20");

        if (options.TryGetValue("status", out var status))
            queryParams["status"] = status;
        if (options.TryGetValue("needs_reply", out var needsReply))
            queryParams["needs_reply"] = needsReply;
        if (options.TryGetValue("assigned_to", out var assignedTo))
            queryParams["assigned_to"] = assignedTo;
        if (options.TryGetValue("channel", out var channel))
            queryParams["channel"] = channel;

        SetupHeaders();
        var response = await httpClient.GetAsync($"{ApiUrl}/conversations?{queryParams}");
        var responseBody = await response.Content.ReadAsStringAsync();

        if (!response.IsSuccessStatusCode)
        {
            throw new Exception($"API Error {(int)response.StatusCode}: {responseBody}");
        }

        return JsonSerializer.Deserialize<JsonElement>(responseBody);
    }

    /// <summary>
    /// Get a single conversation by ID.
    /// </summary>
    private static async Task<JsonElement> GetConversationAsync(string conversationId)
    {
        SetupHeaders();
        var response = await httpClient.GetAsync($"{ApiUrl}/conversations/{conversationId}");
        var responseBody = await response.Content.ReadAsStringAsync();

        if (!response.IsSuccessStatusCode)
        {
            throw new Exception($"API Error {(int)response.StatusCode}: {responseBody}");
        }

        return JsonSerializer.Deserialize<JsonElement>(responseBody);
    }

    /// <summary>
    /// Update a conversation (e.g., assign to a user).
    /// </summary>
    private static async Task<JsonElement> UpdateConversationAsync(string conversationId, string? assignedTo = null)
    {
        var payload = new Dictionary<string, object?>();
        if (assignedTo != null)
        {
            payload["assigned_to"] = assignedTo;
        }

        var json = JsonSerializer.Serialize(payload);
        var content = new StringContent(json, Encoding.UTF8, "application/json");

        SetupHeaders();
        var response = await httpClient.PutAsync($"{ApiUrl}/conversations/{conversationId}", content);
        var responseBody = await response.Content.ReadAsStringAsync();

        if (!response.IsSuccessStatusCode)
        {
            throw new Exception($"API Error {(int)response.StatusCode}: {responseBody}");
        }

        return JsonSerializer.Deserialize<JsonElement>(responseBody);
    }

    /// <summary>
    /// Close a conversation.
    /// </summary>
    private static async Task<JsonElement> CloseConversationAsync(string conversationId)
    {
        SetupHeaders();
        var response = await httpClient.PostAsync($"{ApiUrl}/conversations/{conversationId}/close", null);
        var responseBody = await response.Content.ReadAsStringAsync();

        if (!response.IsSuccessStatusCode)
        {
            throw new Exception($"API Error {(int)response.StatusCode}: {responseBody}");
        }

        return JsonSerializer.Deserialize<JsonElement>(responseBody);
    }

    /// <summary>
    /// Setup common request headers.
    /// </summary>
    private static void SetupHeaders()
    {
        httpClient.DefaultRequestHeaders.Clear();
        httpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", ApiToken);
        httpClient.DefaultRequestHeaders.Add("X-Tenant-ID", TenantId);
    }

    /// <summary>
    /// Print a separator line.
    /// </summary>
    private static void PrintSeparator()
    {
        Console.WriteLine(new string('=', 60));
    }

    /// <summary>
    /// Load environment variables from .env file if it exists.
    /// </summary>
    private static void LoadEnvFile()
    {
        var envPath = Path.Combine(Directory.GetCurrentDirectory(), ".env");
        if (!File.Exists(envPath)) return;

        foreach (var line in File.ReadAllLines(envPath))
        {
            var trimmedLine = line.Trim();
            if (string.IsNullOrEmpty(trimmedLine) || trimmedLine.StartsWith("#")) continue;

            var separatorIndex = trimmedLine.IndexOf('=');
            if (separatorIndex <= 0) continue;

            var key = trimmedLine.Substring(0, separatorIndex).Trim();
            var value = trimmedLine.Substring(separatorIndex + 1).Trim();

            if (string.IsNullOrEmpty(Environment.GetEnvironmentVariable(key)))
            {
                Environment.SetEnvironmentVariable(key, value);
            }
        }
    }
}

File: conversation-management/php/conversation_management.php
<?php
/**
 * SendSeven API - Conversation Management Example (PHP)
 *
 * Demonstrates how to list, get, update, and close conversations using the SendSeven API.
 */

// Load environment variables from .env file (simple implementation)
function loadEnv(string $path): void {
    if (!file_exists($path)) {
        return;
    }
    $lines = file($path, FILE_IGNORE_NEW_LINES | FILE_SKIP_EMPTY_LINES);
    foreach ($lines as $line) {
        if (strpos(trim($line), '#') === 0) {
            continue;
        }
        if (strpos($line, '=') !== false) {
            list($name, $value) = explode('=', $line, 2);
            putenv(trim($name) . '=' . trim($value));
        }
    }
}

loadEnv(__DIR__ . '/.env');

// Configuration from environment
$API_TOKEN = getenv('SENDSEVEN_API_TOKEN');
$TENANT_ID = getenv('SENDSEVEN_TENANT_ID');
$API_URL = getenv('SENDSEVEN_API_URL') ?: 'https://api.sendseven.com/api/v1';

/**
 * Get common headers for API requests.
 */
function getHeaders(): array {
    global $API_TOKEN, $TENANT_ID;
    return [
        'Authorization: Bearer ' . $API_TOKEN,
        'X-Tenant-ID: ' . $TENANT_ID,
        'Content-Type: application/json',
    ];
}

/**
 * List conversations with optional filtering.
 *
 * @param array $options Filter options
 * @return array Paginated list of conversations
 * @throws Exception If the API request fails
 */
function listConversations(array $options = []): array {
    global $API_URL;

    $params = [
        'page' => $options['page'] ?? 1,
        'page_size' => $options['page_size'] ?? 20,
    ];

    if (isset($options['status'])) {
        $params['status'] = $options['status'];
    }
    if (isset($options['needs_reply'])) {
        $params['needs_reply'] = $options['needs_reply'] ? 'true' : 'false';
    }
    if (isset($options['assigned_to'])) {
        $params['assigned_to'] = $options['assigned_to'];
    }
    if (isset($options['channel'])) {
        $params['channel'] = $options['channel'];
    }

    $url = $API_URL . '/conversations?' . http_build_query($params);

    $ch = curl_init($url);
    curl_setopt_array($ch, [
        CURLOPT_RETURNTRANSFER => true,
        CURLOPT_HTTPHEADER => getHeaders(),
    ]);

    $response = curl_exec($ch);
    $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
    $error = curl_error($ch);
    curl_close($ch);

    if ($error) {
        throw new Exception("cURL Error: $error");
    }

    if ($httpCode >= 400) {
        throw new Exception("API Error $httpCode: $response");
    }

    return json_decode($response, true);
}

/**
 * Get a single conversation by ID.
 *
 * @param string $conversationId The UUID of the conversation
 * @return array The conversation object
 * @throws Exception If the API request fails
 */
function getConversation(string $conversationId): array {
    global $API_URL;

    $url = $API_URL . '/conversations/' . $conversationId;

    $ch = curl_init($url);
    curl_setopt_array($ch, [
        CURLOPT_RETURNTRANSFER => true,
        CURLOPT_HTTPHEADER => getHeaders(),
    ]);

    $response = curl_exec($ch);
    $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
    $error = curl_error($ch);
    curl_close($ch);

    if ($error) {
        throw new Exception("cURL Error: $error");
    }

    if ($httpCode >= 400) {
        throw new Exception("API Error $httpCode: $response");
    }

    return json_decode($response, true);
}

/**
 * Update a conversation (e.g., assign to a user).
 *
 * @param string $conversationId The UUID of the conversation
 * @param array $updates Fields to update
 * @return array The updated conversation object
 * @throws Exception If the API request fails
 */
function updateConversation(string $conversationId, array $updates = []): array {
    global $API_URL;

    $url = $API_URL . '/conversations/' . $conversationId;

    $payload = [];
    if (isset($updates['assigned_to'])) {
        $payload['assigned_to'] = $updates['assigned_to'];
    }

    $ch = curl_init($url);
    curl_setopt_array($ch, [
        CURLOPT_CUSTOMREQUEST => 'PUT',
        CURLOPT_POSTFIELDS => json_encode($payload),
        CURLOPT_RETURNTRANSFER => true,
        CURLOPT_HTTPHEADER => getHeaders(),
    ]);

    $response = curl_exec($ch);
    $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
    $error = curl_error($ch);
    curl_close($ch);

    if ($error) {
        throw new Exception("cURL Error: $error");
    }

    if ($httpCode >= 400) {
        throw new Exception("API Error $httpCode: $response");
    }

    return json_decode($response, true);
}

/**
 * Close a conversation.
 *
 * @param string $conversationId The UUID of the conversation
 * @return array The closed conversation object
 * @throws Exception If the API request fails
 */
function closeConversation(string $conversationId): array {
    global $API_URL;

    $url = $API_URL . '/conversations/' . $conversationId . '/close';

    $ch = curl_init($url);
    curl_setopt_array($ch, [
        CURLOPT_POST => true,
        CURLOPT_RETURNTRANSFER => true,
        CURLOPT_HTTPHEADER => getHeaders(),
    ]);

    $response = curl_exec($ch);
    $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
    $error = curl_error($ch);
    curl_close($ch);

    if ($error) {
        throw new Exception("cURL Error: $error");
    }

    if ($httpCode >= 400) {
        throw new Exception("API Error $httpCode: $response");
    }

    return json_decode($response, true);
}

// Main execution
function main(): void {
    global $API_TOKEN, $TENANT_ID;

    // Validate configuration
    if (!$API_TOKEN) {
        echo "Error: SENDSEVEN_API_TOKEN environment variable is required\n";
        exit(1);
    }

    if (!$TENANT_ID) {
        echo "Error: SENDSEVEN_TENANT_ID environment variable is required\n";
        exit(1);
    }

    try {
        // Example 1: List all open conversations that need a reply
        echo str_repeat('=', 60) . "\n";
        echo "Listing open conversations that need a reply...\n";
        echo str_repeat('=', 60) . "\n";

        $result = listConversations([
            'status' => 'open',
            'needs_reply' => true,
            'page_size' => 5,
        ]);

        echo "Found {$result['pagination']['total']} conversations\n";
        echo "Page {$result['pagination']['page']} of {$result['pagination']['total_pages']}\n\n";

        foreach ($result['items'] as $conv) {
            echo "  ID: {$conv['id']}\n";
            echo "  Channel: {$conv['channel']}\n";
            echo "  Status: {$conv['status']}\n";
            echo "  Last message: " . ($conv['last_message_at'] ?? 'N/A') . "\n\n";
        }

        // Example 2: Get a single conversation (if we have any)
        if (!empty($result['items'])) {
            $conversationId = $result['items'][0]['id'];

            echo str_repeat('=', 60) . "\n";
            echo "Getting conversation details: $conversationId\n";
            echo str_repeat('=', 60) . "\n";

            $conversation = getConversation($conversationId);
            echo "  ID: {$conversation['id']}\n";
            echo "  Channel: {$conversation['channel']}\n";
            echo "  Status: {$conversation['status']}\n";
            echo "  Needs reply: " . ($conversation['needs_reply'] ? 'true' : 'false') . "\n";
            echo "  Assigned to: " . ($conversation['assigned_to'] ?? 'Unassigned') . "\n";
            if (isset($conversation['contact'])) {
                echo "  Contact: " . ($conversation['contact']['name'] ?? 'Unknown') . "\n";
            }
            echo "\n";

            // Example 3: Demonstrate update (commented out to avoid modifying data)
            // Uncomment to actually assign a conversation
            // echo str_repeat('=', 60) . "\n";
            // echo "Assigning conversation to user...\n";
            // echo str_repeat('=', 60) . "\n";
            // $userId = 'your-user-id-here';
            // $updated = updateConversation($conversationId, ['assigned_to' => $userId]);
            // echo "  Assigned to: {$updated['assigned_to']}\n\n";

            // Example 4: Demonstrate close (commented out to avoid modifying data)
            // Uncomment to actually close the conversation
            // echo str_repeat('=', 60) . "\n";
            // echo "Closing conversation...\n";
            // echo str_repeat('=', 60) . "\n";
            // $closed = closeConversation($conversationId);
            // echo "  Status: {$closed['status']}\n";
            // echo "  Closed at: {$closed['closed_at']}\n";
        }

        echo str_repeat('=', 60) . "\n";
        echo "Conversation management examples completed!\n";
        echo str_repeat('=', 60) . "\n";

    } catch (Exception $e) {
        echo "Error: " . $e->getMessage() . "\n";
        exit(1);
    }
}

main();
